### PR TITLE
Define simulcast encodings per resolution

### PR DESCRIPTION
**Pull request by Astagor**
This could solve #713 
It is backward compatible with old settings, so it should not break current configs.
@so010 @vpalmisano

### CODE REVIEW

**Comment by vpalmisano** [comment]
I'm not sure that using a scaling factor that is not power of 2 is correct. Have you tried the full hd configuration?

**Comment by Astagor** [comment]
It is OK, you can even use like 3.5 and it works.

**Comment by misi** [comment]
Have you tested it with FF too?

**Comment by Astagor** [comment]
OK, you are right if the values are not the power of 2, FF makes random resolutions. I have changed it.

**Comment by Astagor** [comment]
I did some extensive testing and there is 1 thing I do not unterstand. If I set my camera to low (320x180) and I have in the simulcast
320 :
		[
			{ scaleResolutionDownBy: 1, maxBitRate: 250000 }
		]
this works as expected when it comes to the resolution, but the network quality indicator is "no connection" all the time.

So I have defined a second quality:

320 :
		[
			{ scaleResolutionDownBy: 2, maxBitRate: 100000 },
			{ scaleResolutionDownBy: 1, maxBitRate: 250000 }
		]

BUT with this configuration the browser always sends only 160x90, never 320x180. WHY???

If a cheat a bit and set 2 encodings with scaleResolutionDownBy set to 1 in both
320 :
		[
			{ scaleResolutionDownBy: 1, maxBitRate: 100000 },
			{ scaleResolutionDownBy: 1, maxBitRate: 250000 }
		]

everything works again. I do not get it.

**Comment by so010** [comment]
So lets test this on develop! Thanks @Astagor

**Review by vpalmisano — COMMENTED** [review]
> 320 :
> [
> { scaleResolutionDownBy: 1, maxBitRate: 250000 }
> ]
> this works as expected when it comes to the resolution, but the network quality indicator is "no connection" all the time.

With this configuration at server side from producer stats I get always `score=10`; maybe is there a problem with the UI score indicator?

File: app/src/RoomClient.js [hunk 1]
```diff
@@ -4564,4 +4552,53 @@ export default class RoomClient
...
+	_chooseEncodings(simulcastProfiles, size)
+	{
+		let encodings;
+
+		const sortedMap = new Map([ ...Object.entries(simulcastProfiles) ]
+			.sort((a, b) => parseInt(b[0]) - parseInt(a[0])));
+
+		for (const [ key, value ] of sortedMap)
+		{
+			if (key < size)
```
[the remark this answers is not in the document]
Comment (Astagor) [reply app/src/RoomClient.js:4565]: Hmm, look at the whole for
		for (const [ key, value ] of sortedMap)
		{
			if (key < size)
			{
				if (encodings === null)
				{
					encodings = value;
				}

				break;
			}

			encodings = value;
		}
I'm choosing the one before key<size, so as you say, it is greater then equal.

[the remark this answers is not in the document]
Comment (vpalmisano) [reply app/src/RoomClient.js:4565]: Ok it is the same, thanks.

**Comment by Astagor** [comment]
@vpalmisano OK, so it is GUI based. Also the layers are not displayed, current and preferred, but of course as there is only 1 encoding profile, it is not simulcast anymore, or is it, are there still various temporal layers? Is it allowed to have only 1 encoding profile?

**Comment by vpalmisano** [comment]
IMHO when we have only one level we should disable simulcast.

**Comment by Astagor** [comment]
This means we have to redesign the code, to turn on and off simulcast depending on the number of encoding profiles.
For now for my users I have used the hack:

320 :
[
{ scaleResolutionDownBy: 1, maxBitRate: 100000 },
{ scaleResolutionDownBy: 1, maxBitRate: 250000 }
]

**Comment by Astagor** [comment]
OK, I have made a PR #722 making these changes, so simulcast only gets enabled when there are more then 1 profiles in encodings. However, still the connection UI icon shows "no connection".

**Comment by Astagor** [comment]
To make it more interesting, webcamConsumer.score is undefined when resolution is low or if simulcast is OFF for any resolution... So this is some global problem we have. I did not figure out yet why the score is undefined.

**Comment by Astagor** [comment]
In madiaSoup API 
https://mediasoup.org/documentation/v3/mediasoup/api/#ConsumerScore
they say that the field is required, but I see in the console that it is not there. Am I missing something?
@vpalmisano

**Comment by Astagor** [comment]
OK, it is a race, if there is only 1 video quality the VideoView renders before the consumerScore event comes and that is why it is not set.

**Comment by Astagor** [comment]
OK, I have fixed it in #722 .
But @so010 why in the hell were we doing something like this in server/lib/Room.js
that after sending a 'newConsumer' we were sending the score in a separate notification, instead sending in 'newConsumer'??

**Pull request merged**